pkg/geo/wkt/wkt_test.go

```diff
@@ -334,6 +334,219 @@ func TestUnmarshal(t *testing.T) {
+			},
+			expected: geom.NewGeometryCollection().MustPush(
+				geom.NewGeometryCollection().MustPush(geom.NewLineString(geom.XYM)),
+				geom.NewLineString(geom.XYM),
+			),
+		},
+		{
+			desc: "parse 2D+M geometrycollection with empty geometrycollection",
+			equivInputs: []string{
+				"GEOMETRYCOLLECTION M (GEOMETRYCOLLECTION EMPTY)",
+				"GEOMETRYCOLLECTION M (GEOMETRYCOLLECTION M EMPTY)",
+			},
+			expected: geom.NewGeometryCollection().MustPush(geom.NewGeometryCollection()),
+		},
+		{
+			desc: "parse 3D geometry collection with nested geometrycollection and empty geometry",
+			equivInputs: []string{
+				"GEOMETRYCOLLECTION(GEOMETRYCOLLECTION Z (POLYGON Z EMPTY), POINT Z EMPTY)",
+				"GEOMETRYCOLLECTION(GEOMETRYCOLLECTION(POLYGON Z EMPTY), POINT Z EMPTY)",
+				"GEOMETRYCOLLECTION Z (GEOMETRYCOLLECTION(POLYGON EMPTY), POINT EMPTY)",
+				"GEOMETRYCOLLECTION Z (GEOMETRYCOLLECTION Z (POLYGON EMPTY), POINT EMPTY)",
+				"GEOMETRYCOLLECTION Z (GEOMETRYCOLLECTION Z (POLYGON Z EMPTY), POINT Z EMPTY)",
+			},
+			expected: geom.NewGeometryCollection().MustPush(
+				geom.NewGeometryCollection().MustPush(geom.NewPolygon(geom.XYZ)),
+				geom.NewPointEmpty(geom.XYZ),
+			),
+		},
+		{
+			desc: "parse 2D geometrycollection",
+			equivInputs: []string{`GEOMETRYCOLLECTION(
+POINT(0 0),
+LINESTRING(1 1, 0 0, 1 4),
+POLYGON((0 0, 0 100, 100 100, 100 0, 0 0), (10 10, 11 11, 12 10, 10 10), (2 2, 4 4, 5 1, 2 2)),
+MULTIPOINT((23 24), EMPTY),
+MULTILINESTRING((1 1, 0 0, 1 4)),
+MULTIPOLYGON(((0 0, 0 100, 100 100, 100 0, 0 0))),
+GEOMETRYCOLLECTION EMPTY
+)`},
+			expected: geom.NewGeometryCollection().MustPush(
+				geom.NewPointFlat(geom.XY, []float64{0, 0}),
+				geom.NewLineStringFlat(geom.XY, []float64{1, 1, 0, 0, 1, 4}),
+				geom.NewPolygonFlat(geom.XY,
+					[]float64{0, 0, 0, 100, 100, 100, 100, 0, 0, 0, 10, 10, 11, 11, 12, 10, 10, 10, 2, 2, 4, 4, 5, 1, 2, 2},
+					[]int{10, 18, 26}),
+				geom.NewMultiPointFlat(geom.XY, []float64{23, 24}, geom.NewMultiPointFlatOptionWithEnds([]int{2, 2})),
+				geom.NewMultiLineStringFlat(geom.XY, []float64{1, 1, 0, 0, 1, 4}, []int{6}),
+				geom.NewMultiPolygonFlat(geom.XY, []float64{0, 0, 0, 100, 100, 100, 100, 0, 0, 0}, [][]int{{10}}),
+				geom.NewGeometryCollection(),
+			),
+		},
+		{
+			desc:        "parse 2D geometrycollection with nested geometrycollection",
+			equivInputs: []string{"GEOMETRYCOLLECTION(POINT(0 0), GEOMETRYCOLLECTION(MULTIPOINT(EMPTY, 2 1)))"},
+			expected: geom.NewGeometryCollection().MustPush(
+				geom.NewPointFlat(geom.XY, []float64{0, 0}),
+				geom.NewGeometryCollection().MustPush(
+					geom.NewMultiPointFlat(geom.XY, []float64{2, 1}, geom.NewMultiPointFlatOptionWithEnds([]int{0, 2})),
+				),
+			),
+		},
+		{
+			desc: "parse 2D+M geometrycollection",
+			equivInputs: []string{
+				`GEOMETRYCOLLECTION M (
+POINT EMPTY,
+POINT M (-2 0 0.5),
+LINESTRING M (0 0 200, 0.1 -1 -20),
+POLYGON M ((0 0 7, 1 -1 -50, 2 0 0, 0 0 7)),
+MULTIPOINT M (-1 5 -16, 0.23 7.0 0),
```

Comment:
can you add a test where one of the objects is M when the root is not, and vice versa?

Reply:
Neither of these cases are valid. They're actually in the error cases.

For an XYM geometry collection:
1. All the objects have to use the M variant if the root isn't M.
2. Even if the root is M, all the non-empty geometries have to use M.

Reply:
ah, can you do this for Z?